### PR TITLE
Separate some configurations from '@cocos/build-engine'

### PR DESCRIPTION
**Pull request by shrinktofit**
Re: cocos-creator/3d-tasks#

Changelog:
 * Separate some configurations from build tool `@cocos/build-engine`.

## The reasons

The followings are hard-coded or directly placed in `@cocos/build-engine`:
- how build time constants are generated;
- enumerating modules that do not need transformation;
- enumerating public modules of engine.

Since `@cocos/build-engine` needs to be published onto NPM(otherwise, we get duplicated dependencies of `babel`, `rollup`... with editor), it's hard to make changes when we want to add new platforms, new constants, new dependencies, so on.

This PR however tries to strike this problem.

## Private package `conf`

This PR adds a new **private** module package called `conf` under `/scripts/conf`. "Private" means publishing is never needed and is forbidden(by explicit `private: true` in `package.json`). This package is responsible to reflect some engine meta info.

The one of `conf`'s user is `@cocos/build-engine`. When perform building, `@cocos/build-engine` will by itself locate the package and get the build config contextually(See here https://github.com/cocos-creator/engine/pull/7789/files#diff-839378951bc03176b8a98c8a81097145dc3c0005844d0ec64af1c5ae8d25e1dcR202-R211).

### Setup during development

Just `npm install` & `npm run build` under `/scripts/conf`. This will be included in `npm install` of engine root.

### Setup in release mode

Editor should follow these step to **get a minimal installation**.

Under `/scripts/conf`
1. `npm install` - Perform the development installation; this will also build the package
2. Remove the `node_modules` folder
3. `npm install --only=prod` - Install the production dependencies(e.g. `fs-extra`).

### CODE REVIEW

File: scripts/who-am-i/src/contextual-build-config.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+    switch (context.platform) {
+    default: break;
+    case 'NATIVE':
+        Object.assign(overrides, {
+            'cocos/core/pipeline/index.ts': 'cocos/core/pipeline/index.jsb.ts',
+            'cocos/core/renderer/core/native-pools.ts': 'cocos/core/renderer/core/native-pools.jsb.ts',
+            'cocos/core/gfx/index.ts': 'cocos/core/gfx/index.jsb.ts',
+        });
+        break;
```
Comment (shrinktofit) [inline scripts/who-am-i/src/contextual-build-config.ts]: @minggo  The `moduleOverrides` now moved to here.

File: scripts/who-am-i/src/contextual-build-config.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    };
+
+    return config;
+};
+
+function getTransformExcludes (_context: ConfigContext): NonNullable<Config['transformExcludes']> {
+    return [
+        /node_modules[/\\]@cocos[/\\]ammo/,
+        /node_modules[/\\]@cocos[/\\]cannon/,
+    ];
```
Comment (shrinktofit) [inline scripts/who-am-i/src/contextual-build-config.ts]: Here we can prevent some external dependencies from being transformed. @JayceLai  Notable changes.

File: scripts/who-am-i/src/platform.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    'ALIPAY',
+    'BAIDU',
+    'XIAOMI',
+    'BYTEDANCE',
+    'OPPO',
+    'VIVO',
+    'HUAWEI',
+    'NATIVE',
+    'COCOSPLAY',
+] as const;
```
Comment (shrinktofit) [inline scripts/who-am-i/src/platform.ts]: Platform related. @yanOO1497  @LinYunMo @PPpro

File: scripts/build-engine/src/index.ts [hunk 4]
```diff
@@ -223,165 +196,313 @@ async function _doBuild ({
...
+    const libWhoAmI = ps.join(engineRoot, 'scripts', 'who-am-i', 'lib');
+    const [
+        { getPublicModules, getForceStandaloneModules },
+        { generateBuildTimeConstants },
+        { default: buildConfigFunction },
+    ] = await Promise.all([
+        import(`${libWhoAmI}/public-modules`) as Promise<typeof import('../../who-am-i/lib/public-modules')>,
+        import(`${libWhoAmI}/build-time-constants`) as Promise<typeof import('../../who-am-i/lib/build-time-constants')>,
+        import(`${libWhoAmI}/contextual-build-config`) as Promise<typeof import('../../who-am-i/lib/contextual-build-config')>,
+    ]);
```
Comment (shrinktofit) [inline scripts/build-engine/src/index.ts]: The connection.

File: scripts/build-engine/src/index.ts [hunk 5]
```diff
@@ -223,165 +196,313 @@ async function _doBuild ({
...
+export default ammo.default;
+export { isWasm };
+`;
+    } else if (ammoJsWasm === true) {
+        rpVirtualOptions['@cocos/ammo'] = `
+import Ammo from '${filePathToModuleRequest(ammoJsWasmModule)}';
+export default Ammo;
+const isWasm = false;
+export { isWasm };
+`;
```
Comment (shrinktofit) [inline scripts/build-engine/src/index.ts:255]: In future I'm going to generalize the code here. But I have no idea now. @JayceLai

File: scripts/who-am-i/src/contextual-build-config.ts (hunk 1 above)
Comment (minggo) [reply scripts/who-am-i/src/contextual-build-config.ts]: I think you missed this one 
```
"cocos/core/gfx/pipeline-state.ts": "cocos/core/gfx/pipeline-state.jsb.ts"
```

And what should i do to test it? Should i run `npm install` after merging this PR?

Comment (shrinktofit) [reply scripts/who-am-i/src/contextual-build-config.ts]: Yes. Just `npm install`

File: scripts/who-am-i/src/platform.ts (hunk 3 above)
Comment (PPpro) [reply scripts/who-am-i/src/platform.ts]: need 
QTT and LINKSURE

File: scripts/who-am-i/src/build-time-constants.ts [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    knownPlatforms.forEach((name) => constants[name] = false);
+    if (platform) {
+        if (knownPlatforms.includes(platform as KnownPlatform)) {
+            console.warn(`Unknown platform constant: ${platform}`);
+        } else {
+            console.warn(`Unknown platform constant: ${platform}`);
+        }
+    }
+
+    constants.DEV = constants.EDITOR || constants.PREVIEW || constants.TEST;
```
Comment (yanOO1497) [inline scripts/who-am-i/src/build-time-constants.ts]: 以前这种宏的配置模式，新增平台就需要在这里修改代码逻辑，希望达到的效果是，外部有 json 对针对平台做这些宏的配置。比如 

```json5
{
    Plaform: 'OPPO',
    constants: {
        SUPPORT_JIT: false,
        JSB: false,
        ......
    }
}

```

对于引擎而言是不需要知道平台概念的，只需要知道设置对应宏的值即可。这些根据不同平台发生变化的宏配置，应该开放出去让平台自己配置。

**Pull request closed without merging**